zerver/actions/uploads.py

```diff
@@ -85,40 +88,56 @@ def do_delete_old_unclaimed_attachments(weeks_ago: int) -> None:
+        archived_attachment.path_id for archived_attachment in old_unclaimed_archived_attachments
+    ]
+
+    ImageAttachment.objects.filter(
+        Q(path_id__in=attachments_path_ids) | Q(path_id__in=archived_attachments_path_ids)
+    ).update(deleted=True)
+
     # An attachment may be removed from Attachments and
     # ArchiveAttachments in the same run; prevent warnings from the
     # backing store by only removing it from there once.
     already_removed = set()
     storage_paths = []
-    for attachment in old_unclaimed_attachments:
-        storage_paths.append(attachment.path_id)
-        image_row = ImageAttachment.objects.filter(path_id=attachment.path_id).first()
+    for attachment_path_id in attachments_path_ids:
+        storage_paths.append(attachment_path_id)
+        image_row = ImageAttachment.objects.filter(path_id=attachment_path_id).first()
         if image_row:
             for existing_thumbnail in image_row.thumbnail_metadata:
                 thumb = StoredThumbnailFormat(**existing_thumbnail)
                 storage_paths.append(get_image_thumbnail_path(image_row, thumb))
-            image_row.delete()
-        already_removed.add(attachment.path_id)
-        attachment.delete()
-        if len(storage_paths) >= DELETE_BATCH_SIZE:
-            delete_message_attachments(storage_paths[:DELETE_BATCH_SIZE])
-            storage_paths = storage_paths[DELETE_BATCH_SIZE:]
-    for archived_attachment in old_unclaimed_archived_attachments:
-        if archived_attachment.path_id not in already_removed:
-            storage_paths.append(archived_attachment.path_id)
-            image_row = ImageAttachment.objects.filter(path_id=archived_attachment.path_id).first()
+        already_removed.add(attachment_path_id)
+
+    for archived_attachment_path_id in archived_attachments_path_ids:
+        if archived_attachment_path_id not in already_removed:
+            storage_paths.append(archived_attachment_path_id)
+            image_row = ImageAttachment.objects.filter(path_id=archived_attachment_path_id).first()
             if image_row:  # nocoverage
                 for existing_thumbnail in image_row.thumbnail_metadata:
                     thumb = StoredThumbnailFormat(**existing_thumbnail)
                     storage_paths.append(get_image_thumbnail_path(image_row, thumb))
-                image_row.delete()
-        archived_attachment.delete()
-        if len(storage_paths) >= DELETE_BATCH_SIZE:
-            delete_message_attachments(storage_paths[:DELETE_BATCH_SIZE])
-            storage_paths = storage_paths[DELETE_BATCH_SIZE:]
-
-    if storage_paths:
-        delete_message_attachments(storage_paths)
+
+    # delete DELETE_BATCH_SIZE files one at a time, from storage
+    for batch_start_index in range(0, len(storage_paths), DELETE_BATCH_SIZE):
+        delete_message_attachments(
+            storage_paths[batch_start_index : batch_start_index + DELETE_BATCH_SIZE]
+        )
```

Comment:
Does `delete_message_attachments` continue properly in both backends if some of the paths don't exist?

Reply:
good question !
## for local backend
it continues yes, and just raises this warning
```python
logging.warning("%s does not exist. Its entry in the database will be removed.", file_name)
```
The way I tested this is by making an early deletion for a file using `delete_local_file` before calling `delete_message_attachments`
## for S3 backend

I think yes, function that's used is [S3UploadBackend.delete_message_attachments](https://github.com/zulip/zulip/blob/main/zerver/lib/upload/s3.py#L286-L289)

but there are no warnings for non-existing files, because [Note that if the object specified in the request is not found, Amazon S3 returns the result as deleted.](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3/bucket/delete_objects.html#:~:text=Note%20that%20if%20the%20object%20specified%20in%20the%20request%20is%20not%20found%2C%20Amazon%20S3%20returns%20the%20result%20as%20deleted.)